### PR TITLE
update load interited widget code lab

### PR DESCRIPTION
**Pull request by chunhtai**

### CODE REVIEW

**Comment by chunhtai** [comment]
cc @johnpryan

**Comment by domesticmouse** [comment]
Please update to push CI green on beta and dev channels

**Comment by chunhtai** [comment]
blocked on https://github.com/flutter/codelabs/pull/92

**Comment by domesticmouse** [comment]
Can this PR be made not to block on #92? My understanding (@Piinks to correct me if I'm wrong) is that #92 is blocked until Flutter 2.2 is launched, which is I'm guessing is unacceptable for this codelab.

**Comment by Piinks** [comment]
I am not sure how this is blocked by #92, AFAICT that is a separate codelab, and not used by this codelab. @chunhtai can you explain how you are blocked?

I was wondering for #97 if it is blocked because it relies on code that has not reached stable yet, is it similar here?

**Comment by chunhtai** [comment]
I was hitting the build failure about the drag and drop deprecation warning and thought i have to wait for https://github.com/flutter/codelabs/pull/92. It looks like the build is happy somehow, so i think this is not blocked then.

**Comment by domesticmouse** [comment]
The code in this PR isn't opted into CI, so it shouldn't be blocked from using features not yet in stable. 

We had some of the codelabs die when `beta` and `dev` channels were updated a week back, which was blocking all PRs.

**Comment by chunhtai** [comment]
> The code in this PR isn't opted into CI, so it shouldn't be blocked from using features not yet in stable.
> 
> We had some of the codelabs die when `beta` and `dev` channels were updated a week back, which was blocking all PRs.

That is probably it then. I think this pr is ready for review now.

**Review by domesticmouse — APPROVED** [review]
LGTM, but please wait for @johnpryan

File: dartpad_codelabs/src/inherited_widget/intro/instructions.md [hunk 1]
```diff
@@ -0,0 +1,31 @@
+# Introduction
+
+State management has been a really important topic in mobile development.
+In Flutter, We provide the InheritedWidget aim to make the process of building
```
Comment (johnpryan) [inline dartpad_codelabs/src/inherited_widget/intro/instructions.md:4]:
```suggestion
In Flutter, InheritedWidget makes the process of building
```

File: dartpad_codelabs/src/inherited_widget/intro/instructions.md [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+the state management system easier for our developers. In this workshop,
```
Comment (johnpryan) [inline dartpad_codelabs/src/inherited_widget/intro/instructions.md:5]:
```suggestion
the state management system easier. In this workshop,
```

File: dartpad_codelabs/src/inherited_widget/intro/instructions.md [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+you will learn how to handle your app state by using the InheritedWidget.
```
Comment (johnpryan) [inline dartpad_codelabs/src/inherited_widget/intro/instructions.md:6]:
```suggestion
you will learn how to handle your app state by using InheritedWidget.
```

File: dartpad_codelabs/src/inherited_widget/intro/instructions.md [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+
+
+
+# An Overview
+
+This workshop includes a series of steps to convert an app without
+any state management to a fully centralized state management system.
+
+You will be started with a demo Google store app. To see how the app looks
+like, click the `Run` button on the top right corner of the IDE.
```
Comment (johnpryan) [inline dartpad_codelabs/src/inherited_widget/intro/instructions.md:17]:
```suggestion
You will be started with a demo Google store app. To see how the app looks,
click the `Run` button in the top right corner of the IDE.
```

File: dartpad_codelabs/src/inherited_widget/intro/instructions.md [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+
+Try to scroll through different google products and add/remove products from
+the cart. The small cart icon on the right of the app bar changes based on number
+of the products in the cart. You can also perform search queries by clicking the
+search icon in the app bar and type `"nest"` to filter nest products.
+
+There are two components that require states:
```
Comment (johnpryan) [inline dartpad_codelabs/src/inherited_widget/intro/instructions.md:24]:
```suggestion
There are two components that require stateful objects:
```

File: dartpad_codelabs/src/inherited_widget/intro/snippet.dart [hunk 6]
```diff
@@ -0,0 +1,327 @@
...
+  );
+}
+
+class MySorePage extends StatefulWidget {
+  MySorePage({Key? key}) : super(key: key);
+  @override
+  MySorePageState createState() => MySorePageState();
+}
+
+class MySorePageState extends State<MySorePage> {
```
Comment (johnpryan) [inline dartpad_codelabs/src/inherited_widget/intro/snippet.dart:109]: `MyStorePageState`

File: dartpad_codelabs/src/inherited_widget/intro/snippet.dart [hunk 7]
```diff
@@ -0,0 +1,327 @@
+// Copyright (c) 2019, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'package:flutter/widgets.dart';
+import 'package:flutter/material.dart';
+
+const String baseAssetURL = 'http://localhost:8000/example/codelabs/inherited_widget/assets';
```
Comment (johnpryan) [inline dartpad_codelabs/src/inherited_widget/intro/snippet.dart:8]: 👍 - We'll make sure to change this to the hosted URL when we have one

File: dartpad_codelabs/src/inherited_widget/step_01/instructions.md [hunk 8]
```diff
@@ -0,0 +1,72 @@
+# Define the State Data Structure
+
+The first thing to do is defining a data structure to store your app state.
```
Comment (johnpryan) [inline dartpad_codelabs/src/inherited_widget/step_01/instructions.md:3]:
```suggestion
The first thing to do is define a data structure to store your app state.
```

File: dartpad_codelabs/src/inherited_widget/step_01/instructions.md [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+To do that, we need to find what should be considered as the app state.
```
Comment (johnpryan) [inline dartpad_codelabs/src/inherited_widget/step_01/instructions.md:4]:
```suggestion
To do that, we need to determine what should be considered the app state.
```

File: dartpad_codelabs/src/inherited_widget/step_01/instructions.md [hunk 10]
```diff
@@ -0,0 +1,72 @@
...
+
+We know there are at least two components requires state, the product list
```
Comment (johnpryan) [inline dartpad_codelabs/src/inherited_widget/step_01/instructions.md:6]:
```suggestion
We know there are at least two components require state, the product list
```

File: dartpad_codelabs/src/inherited_widget/step_01/solution.dart [hunk 11]
```diff
@@ -0,0 +1,337 @@
...
+      _purchaseList = value;
+    });
+  }
+
+  @override
+  Widget build(BuildContext context) {
+    final bool hasPurchase = purchaseList.length > 0;
+    return Stack(
+      alignment: Alignment.center,
+      children: <Widget>[
```
Comment (johnpryan) [inline dartpad_codelabs/src/inherited_widget/step_01/solution.dart:198]: nitpick:

```suggestion
      children: [
```

File: dartpad_codelabs/src/inherited_widget/step_02/instructions.md [hunk 12]
```diff
@@ -0,0 +1,25 @@
+# Create a InheritedWidget
+
+We have our own data structure to store the states. The next thing we want
```
Comment (johnpryan) [inline dartpad_codelabs/src/inherited_widget/step_02/instructions.md:3]:
```suggestion
Now we have our own data structure to store the states. The next thing we want
```

File: dartpad_codelabs/src/inherited_widget/step_03/instructions.md [hunk 13]
```diff
@@ -0,0 +1,40 @@
+# Create a StatefulWidget around the AppStateScope
+
+The `AppStateScope` itself does not have state, it only hosts the data
+that it receives. We still need to create a `StatefulWidget` to store
+the data. The goal of this `StatefulWidget` is to create the `StateData`
+, provide APIs to mutate the data and host the data using the
```
Comment (johnpryan) [inline dartpad_codelabs/src/inherited_widget/step_03/instructions.md:6]: nit: oxford comma:
```suggestion
, provide APIs to mutate the data, and host the data using the
```

File: dartpad_codelabs/src/inherited_widget/intro/instructions.md [hunk 14]
```diff
@@ -0,0 +1,31 @@
+# Introduction
+
+State management has been a really important topic in mobile development.
```
Comment (johnpryan) [inline dartpad_codelabs/src/inherited_widget/intro/instructions.md:3]:
```suggestion
State management is an important topic in mobile development.
```

File: dartpad_codelabs/src/inherited_widget/step_05/instructions.md [hunk 15]
```diff
@@ -0,0 +1,23 @@
+# Migrate the ShoppingCartIcon
+
+Every widget should have access to the `AppStateWidget` and `AppStateScope`.
+Let's migrate `ShoppingCartIcon` to use the `AppStateScope` instead of storing
+the state locally.
+
+(Note# The app will not work properly until we migrate all of the widgets.)
```
Comment (johnpryan) [inline dartpad_codelabs/src/inherited_widget/step_05/instructions.md:7]: I'm not sure if we will support note syntax, consider just using bold text:
```suggestion
**Note:** The app will not work properly until we migrate all of the widgets.
```

File: dartpad_codelabs/src/inherited_widget/step_06/instructions.md [hunk 16]
```diff
@@ -0,0 +1,17 @@
...
+  ProductListWidget({Key? key}) : super(key: key);
+  @override
+  ProductListWidgetState createState() => ProductListWidgetState();
+}
+```
+
+Remove the global key usage.
+
+```dart
+// TODO: remove the usage of productList globalkey.
```
Comment (johnpryan) [inline dartpad_codelabs/src/inherited_widget/step_06/instructions.md:15]: We should probably change the comment to use the GlobalKey class name:

```suggestion
// TODO: remove the usage of productList GlobalKay.
```

File: dartpad_codelabs/src/inherited_widget/step_06/snippet.dart [hunk 17]
```diff
@@ -0,0 +1,391 @@
...
+    required this.title,
+    required this.description
+  });
+  final String id;
+  final String pictureURL;
+  final String title;
+  final TextSpan description;
+}
+
+// TODO: remove the usage of productList globalkey.
```
Comment (johnpryan) [inline dartpad_codelabs/src/inherited_widget/step_06/snippet.dart:90]:
```suggestion
// TODO: remove the usage of productList GlobalKey.
```

File: dartpad_codelabs/src/inherited_widget/step_08/instructions.md [hunk 18]
```diff
@@ -0,0 +1,18 @@
+# Migrate MySorePage
+
+Now we have successfully migrate the app to use a centralized state management system.
```
Comment (johnpryan) [inline dartpad_codelabs/src/inherited_widget/step_08/instructions.md:3]:
```suggestion
Now we have successfully migrated the app to use a centralized state management system.
```

File: dartpad_codelabs/src/inherited_widget/step_08/instructions.md [hunk 19]
```diff
@@ -0,0 +1,18 @@
...
+However, there is still one StatefulWidget that contains its own state. That is
+`MySorePage` widget.
```
Comment (johnpryan) [inline dartpad_codelabs/src/inherited_widget/step_08/instructions.md:5]:
```suggestion
`MyStorePage` widget.
```

File: dartpad_codelabs/src/inherited_widget/step_08/instructions.md [hunk 20]
```diff
@@ -0,0 +1,18 @@
...
+
+This widget keeps track of whether it is in a search mode and `TextField` related
+controllers. For the purpose of this tutorial, let's see if we can completely eliminate
+any state hanging around the app.
+
+```dart
+// TODO: convert MySorePage into StatelessWidget
+class MySorePage extends StatefulWidget {
```
Comment (johnpryan) [inline dartpad_codelabs/src/inherited_widget/step_08/instructions.md:13]:
```suggestion
// TODO: convert MyStorePage into StatelessWidget
class MyStorePage extends StatefulWidget {
```

File: dartpad_codelabs/src/inherited_widget/step_08/instructions.md [hunk 21]
```diff
@@ -0,0 +1,18 @@
...
+  MySorePage({Key? key}) : super(key: key);
+  @override
+  MySorePageState createState() => MySorePageState();
```
Comment (johnpryan) [inline dartpad_codelabs/src/inherited_widget/step_08/instructions.md:16]:
```suggestion
  MyStorePage({Key? key}) : super(key: key);
  @override
  MyStorePageState createState() => MySorePageState();
```

File: dartpad_codelabs/src/inherited_widget/step_08/snippet.dart [hunk 22]
```diff
@@ -0,0 +1,373 @@
...
+  @override
+  Widget build(BuildContext context) {
+    return AppStateScope(
+      _data,
+      child: widget.child,
+    );
+  }
+}
+
+// TODO: convert MySorePage into StatelessWidget
```
Comment (johnpryan) [inline dartpad_codelabs/src/inherited_widget/step_08/snippet.dart:176]:
```suggestion
// TODO: convert MyStorePage into StatelessWidget
```

File: dartpad_codelabs/src/inherited_widget/step_09/instructions.md [hunk 23]
```diff
@@ -0,0 +1,4 @@
+# Thank you for participating in this workshop
+
+Congratulation! We have reach the end of this workshop!
```
Comment (johnpryan) [inline dartpad_codelabs/src/inherited_widget/step_09/instructions.md:3]:
```suggestion
Congratulations! You've successfully completed this workshop!
```

File: dartpad_codelabs/src/inherited_widget/step_09/instructions.md [hunk 24]
```diff
@@ -0,0 +1,4 @@
...
+
```
Comment (johnpryan) [inline dartpad_codelabs/src/inherited_widget/step_09/instructions.md:4]: We should consider adding some links to our API docs to learn more about InheritedWidget and other relevant APIs

File: dartpad_codelabs/src/inherited_widget/step_08/instructions.md (hunk 19 above)
Comment (johnpryan) [reply dartpad_codelabs/src/inherited_widget/step_08/instructions.md:5]: We need to update this everywhere, so I didn't add a suggestion for every place this is used.

**Comment by johnpryan** [comment]
I'm going to merge and deploy this - our tech writers will review the staged version.

**Pull request merged**